common/src/main/scala/ClassTwoAndThreeResult.scala

```diff
@@ -18,32 +18,74 @@ package eoi
 
 import spire.math.Interval
 import cats.syntax.applicative._
-import cats.syntax.traverse._
-import cats.instances.list._
 import cats.syntax.apply._
 import spire.implicits._
 import java.time.LocalDate
 
+trait ClassTwoOrThree {
+  def noOfWeeks: Int
+  def rate: BigDecimal
+  def lowerEarningLimit: Explained[BigDecimal]
+  def qualifyingEarningsFactor: Explained[BigDecimal]
+  def finalDate: Option[LocalDate]
+}
+
+case class ClassTwo(
+  weeklyRate: ClassTwoRates,
+  smallEarningsException: Option[BigDecimal],
+  hrpDate: Option[LocalDate],
+  finalDate: Option[LocalDate],
+  noOfWeeks: Int = 52,
+  qualifyingRate: BigDecimal
+) extends ClassTwoOrThree {
+  def rate = weeklyRate.default
+  def lowerEarningLimit = ((qualifyingRate + 50) / noOfWeeks) gives
+    s"lowerEarningLimit: (qualifyingRate + 50) / noOfWeeks = ($qualifyingRate + 50) / $noOfWeeks"
+
+  def qualifyingEarningsFactor: Explained[BigDecimal] = qualifyingRate.pure[Explained]
+}
+
+case class ClassThree(
+  finalDate: Option[LocalDate],
+  weekRate: BigDecimal,
+  noOfWeeks: Int = 52,
+  lel: BigDecimal,
+  qualifyingRate: Option[BigDecimal]
+) extends ClassTwoOrThree {
+  def rate = weekRate
+  def qualifyingEarningsFactor: Explained[BigDecimal] = qualifyingRate match {
```

Comment:
isn't the qualifying earnings factor always in config? If it is then couldn't we just take it from there rather than calculate it?

Reply:
Do you not think it makes sense to have the fallback? Given that we know how it is calculated?

Reply:
it would make sense for now while the rules apply - if the rules ever change in subsequent years then this piece of code would have to be updated, whereas if the QEF is always taken out of config then it's the responsibility of whoever is doing the update to ensure the right figure is entered. Does any of the existing config has QEF missing?

Reply:
I think they all have the QEF defined. Though when I parse the 2020 db I might only put it in explicitly if it differs from what is computed. 
The same thing applied with the final date at present - in most cases that is computed but in a few historical cases it differs. 

I'm thinking 'convention over configuration' here.